private-templates-service/src/controllers/template.ts

```diff
@@ -1,19 +1,78 @@
 import passport from '../config/passport';
 import express from 'express';
+import { Request, Response, NextFunction } from 'express';
+import { Template } from '../models/Template';
+import mongoose from 'mongoose';
+import logger from '../util/logger';
+import jws from 'jws';
+import { check, validationResult } from 'express-validator';
 
 const templateRouter = express.Router();
 
 templateRouter.all('/', passport.authenticate('oauth-bearer', {
     session: false
-    }), function(req: any, res: any, next: Function){
+    }), function(req: Request, res: Response, next: NextFunction){
     next();
 });
 
-// create a GET route
-templateRouter.get('/', (req : any, res : any) => {
-    res.send({
-        express: "YOUR EXPRESS BACKEND IS CONNECTED TO REACT. CONFIRMED"
+// GET templates route
+const getTemplates = (req: Request, res: Response, next: NextFunction) => {
+    Template.find((err, templates) => {
+        if (err) return next(err);
+        res.json(templates);
     });
-});
+}
+
+const getTemplateById = (req: Request, res: Response, next: NextFunction) => {
+    const templateId = req.params.id;
+    Template.findById(templateId, (err, template) => {
+        if (err) return res.status(404).send('Template does not exist.');
+        res.json(template);
+    })
+}
+
+// POST templates route
+const postTemplates = async(req: Request, res: Response, next: NextFunction) => {
+    // TODO: add more checks to validate template
+    await check('template', 'Template is not valid JSON.').isJSON().run(req);
+
+    const errors = validationResult(req);
+
+    if (!errors.isEmpty()) {
+        logger.error(errors);
+        return res.status(400).json(errors);
+    }
+
+    // Retrieve bearer token & decode oid
+    var bearer = req.get('Authorization');
+    if (!bearer) {
+        return res.status(400).send('No access token was found.');
+    }
+    bearer = bearer.split(/[ ]+/).pop();
+    if (!bearer) {
+        return res.status(400).send('Malformatted access token.');
+    }
+
+    var decodedTokenOid = jws.decode(bearer).payload.oid;
+    logger.info('Owner is: ' + decodedTokenOid);
+
+    const template = new Template({
```

Comment:
Do you have a plan for extracting this logic to it's own package in future?  The sort of structure Matt was after (from what I understand) is that business logic like this will be in it's own package.  

The code here would then look something like:

import adaptivecarservice from 'adaptivecardsservice'
cardService = new adaptiveCardService(.......)

Then in places like this just:
cardService.saveCard(template)

This would allow other people to use all the business logic in their own web service if they want (and then the API here would just wrap the logic and provide authentication)

Reply:
I'm not 100% sure how we would extract this logic right now, I'll look into it and follow up!

Reply:
Cool, just something to keep in mind when writing this stuff